[24.0] Fix Multi-History status bar reactivity

davelopez opened the pull request:
Fixes reactivity issues related to the `HistoryCounter`, `DefaultOperations`, and the History store.

Adds stronger typing to history-related components, especially the HistoryStore, to clarify which information is being stored and retrieved when dealing with histories. This includes a new schema model `HistoryDevDetailed` used by the "client controller API" and adds more missing fields. Although our FastAPI endpoints do not use this model, it contains the information we are working with when requesting history details in the UI and it's slightly different from the `HistoryDetailed` model.

## TODO
- [ ] <del>Isolate the target history from the "current history" when operating on histories in the multi-view.</del> We cannot really do this without creating or modifying endpoints to support multi-histories and "monitor" the state of those.
- [x] Fixes #17697 by temporarily disabling any operation (except for selecting and dragging/dropping items into histories) on histories in the multi-view. This should be properly fixed in `dev` by fully supporting multi-histories (see first point).
- [ ] Fixes #17698

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Refactor async calls to avoid mixing `await` and `then`

davelopez committed:
Reduce redundant detailed history load

The current history details will be loaded by the history watcher anyway.

davelopez committed:
Add HistoryDevDetailed schema model

Technically this model is used by the "client API" in the old controllers so the current API is not using it, but it is useful to have the schema in the client to know what we are getting when requesting something from the server.

davelopez committed:
Load current history through the historyStore in watcher

This will make the history data available for all the app and reduce duplicated requests.

davelopez committed:
Use proper types for history requests in the store

This should give us an idea of what information is
being stored and retrieved each time.

davelopez committed:
Add history placeholder to prevent multiple requests

And avoid reactivity issues with lost references

davelopez committed:
Fix multi-history view stats

Add composables to deal with "extended" versions of histories when needed.
Use proper types for history references in components.

davelopez committed:
Refactor HistoryScrollList

To use AnyHistory type and userOwnsHistory function

davelopez committed:
Refactor MultipleView.test.js to improve test setup

davelopez committed:
Drop history placeholder as it doesn't really help

davelopez committed:
Refactor useExtendedHistory composable

davelopez committed:
Add updateContentStats action to histories store

This allows to request only those fields that might have changed after an operation in the history contents.

davelopez committed:
Disable history operations conditional if multi-view

I don't like this at all, but we don't have a way to deal with those operations in a multi-view scenario. This should be a temporary fix util we work on enabling real multi-history support both in backend and client.

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -1390,21 +1410,29 @@ class HistoryDetailed(HistorySummary):  # Equivalent to 'dev-detailed' view, whi
+    Currently used by the `HistoryController` endpoints.
+    """
+
+    contents_active: HistoryActiveContentCounts = Field(
```
Could you request this separately instead, like the history grid does ? (e.g. https://usegalaxy.org/api/histories/18173ac7e98c6e7a?keys=nice_size,contents_active,contents_states) ?

davelopez replied:
This model is only used by the `history/current_history_json` (not the regular API) and I'm using it right now just to know what fields we dealing with in the store.

mvdbeek replied:
Sounds good, but let's not add it to the AnyHistoryView then ?

davelopez replied:
That is a good point, I should have mentioned that, unfortunately, if the model is not referenced in any API route, then, the client schema, does not include/generate it :disappointed: 

In theory, this dev-detailed model is a "valid" view that we can request with `view=dev-detailed` https://github.com/galaxyproject/galaxy/blob/6b96bb17b2099c527fd5e6b02699e8f6b81b6593/lib/galaxy/managers/histories.py#L816 so doesn't seem to hurt to put it in the AnyHistoryView for now.

We should work on `dev` on migrating this `history/current_history_json` to a regular API endpoint that also supports other histories apart from the current one and maybe that would be the perfect time to fine-tune which fields we should really serialize and then request the expensive ones on demand from the client. All this with proper strong types both in the backend and frontend.

mvdbeek replied:
Is there maybe a lighter-weight way to handle this ? I dislike adding types to our API that aren't actually part of our API ... could you take `CustomHistoryView` and pick the fields in your type declaration ?

davelopez replied:
Yes I could just declare the interface in the client and drop it from the "AnyHistoryView" it is something temporal anyway :+1:

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -1331,42 +1369,24 @@ class HistoryActiveContentCounts(Model):
     )
-    create_time: datetime = CreateTimeField
-    importable: bool = Field(
+    empty: bool = Field(
```
This seems wrong. This is quite redundant with the hid_counter or whatever we call it.

davelopez replied:
I agree, there are a bunch of redundant properties also like `nice_size` or `type_id` (both can be computed in the client) or `file_ext` and `extension` for datasets.

This `empty` seems to have been there since 2014 https://github.com/galaxyproject/galaxy/blame/6b96bb17b2099c527fd5e6b02699e8f6b81b6593/lib/galaxy/managers/histories.py#L793 on the other hand, what we really are using for detailed views in the client is this `dev-detailed` view (and not the `detailed`) because most of the requests for details are for the current history with `history/current_history_json`.

Again we should really try to untangle this serialization and fine-tune the models.

Should we do something about this in this PR?

mvdbeek replied:
> because most of the requests for details are for the current history with history/current_history_json.

what a mess :(. 

> Should we do something about this in this PR?

can we get away with not changing the models ?

davelopez replied:
Yes, sure! I already felt there were too many changes for a bug fix, but I wanted to "accurately expose" what we are "really" retrieving from the server and storing in the client.

I can revert the API model changes in the backend and try to mimic them in the client to achieve something similar. Then we can follow up on dev trying to migrate this `history/current_history_json` and associates to the API and fine-tune the models at that point if we can.

Sounds good?

mvdbeek replied:
Yes, thank you!

davelopez replied:
Done! I've documented in the client schemas `client/src/api/index.ts` the inconsistencies with the views and what we usually request from the client and serialize from the backend.

davelopez committed:
Revert schema changes in backend

Let's do this on dev, after we evaluate what to improve.

davelopez committed:
Adapt client API schemas for histories

To accurately reflect what the server is serializing for each different view or usual request from the client.

dannon committed:
Address reactivity bug; ensure newly added properties are reactive -- this is much cleaner in vue3...

dannon reviewed (approved):
Thanks @davelopez!

nsoranzo commented on client/src/components/History/Multiple/MultipleView.test.js:
```diff
@@ -41,21 +51,22 @@ describe("MultipleView", () => {
-        const { wrapper, axiosMock } = await setUpWrapper(UserHistoriesMock, count, currentHistoryId);
+        const wrapper = await setUpWrapper(count, currentHistoryId);
+
+        console.log(wrapper.html());
```
Debug leftover?

davelopez replied:
absolutely... :sweat_smile: Thanks!

:eagle: :eye: :smile:
